dot/state/roottottrie.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
+
+func newRootToTrieMap() *rootToTrieMap {
+	return &rootToTrieMap{
+		rootToTrie: make(map[common.Hash]*trie.Trie),
+	}
+}
+
+func (r *rootToTrieMap) get(root common.Hash) (t *trie.Trie, has bool) {
+	r.RLock()
+	defer r.RUnlock()
+	t, has = r.rootToTrie[root]
+	return t, has
+}
+
+func (r *rootToTrieMap) set(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) setIfUnset(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+
+	_, exists := r.rootToTrie[root]
+	if exists {
+		return
+	}
+
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) delete(root common.Hash) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.rootToTrie, root)
+}
+
+func (r *rootToTrieMap) getFromMemOrDB(root common.Hash, db chaindb.Database) (
+	t *trie.Trie, err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	var calculatedRoot common.Hash
+	t, has := r.rootToTrie[root]
+	if has {
+		calculatedRootHash := t.MustHash()
+		if !calculatedRootHash.Equal(root) {
+			panic(fmt.Sprintf("trie does not have expected root, expected %s but got %s",
+				root, calculatedRootHash))
+		}
```

Comment:
Can I remove those panics (here and below)? Or is there a risk of a root hash mismatch here? :thinking: 

That might accelerate processing since we don't need to encode and hash the entire trie (`t.MustHash()`)

Reply:
yeah, you can remove them, in the case where the trie gets corrupted for whatever reason and the roots don't match, the block processing higher up will fail with a state root mismatch

Reply:
So I removed the check when loading from memory since it's rather not possible to have it corrupt I think. I left the one when loading the trie from disk.

By the way, I ran gossamer with a print to see how often it would load the trie from disk in this function and it looks like it never does (for 30 min), is this expected?

Reply:
yeah, it shouldn't really ever load from disk except on start-up, so that sounds about right

Reply:
it's rather slow to load from disk so we don't want that when syncing

Reply:
Ok that makes sense. Maybe some simplification could be done, since it likely loads from disk from another code path so this code path in this function is never really run. Anyway, in another PR later maybe some day in the future.

Reply:
I changed it to be used in more places, now the read from disk code path is indeed called. 👍 
That way I also spotted a child trie test bug with the root hash panic check when reading from disk, although I'm still working on debugging it.

Reply:
Root hash check moved to PR #2251 (with child trie test failing for now)